design-proposals/mass_machine_type_transition.md

```diff
@@ -0,0 +1,299 @@
+the user to automatically update the machine type of any VMs with a machine type
+that is no longer supported. This command invokes a k8s `Job` that iterates
+through all VMs within a specified namespace (or all namespaces if none is
+specified), determines if the machine type is no longer supported, and updates
+it to the latest supported version. If the Job fails or is killed, it will be
+restarted with `restartPolicy=OnFailure` in the Job spec. The `restart-vm-required`
+label may then remain on a VM that has already been shut down, so the Job will
+first check VMs with the label and confirm if the machine types have been updated.
+If they have, then the label will be removed, but if not, the label will remain
+and the VM will continue to be tracked until it is restarted and its machine
+type is updated. Both the minimum supported machine type and the target machine
+type will be determined (retrieved) and configured internally when the command is
+invoked, and these values will automatically be updated as future versions are
+released and more machine types are deprecated upon these releases.
+
+If the VM is running, the update will not take effect until it is restarted,
+so the label `restart-vm-required` will be applied to the VM until the VMI is
+removed (indicating the VM has been stopped), upon which the label will be cleared
+after verifying that its machine type has been successfully updated. The job will
+only terminate once all `restart-vm-required` labels have been cleared. By default,
+the user must handle stopping/restarting the labelled VMs manually. This is to allow
+the user to safely update the machine types of their VMs at a time they choose
+without worrying about workload interruptions. However, the user also has the
+option to allow the job to automatically restart every running VM that is being
+updated and apply the changes immediately.
+
+The options to specify by namespace, label-selector, or restart the VMs
+immediately are configurable with the `--namespace`, `--label-selector`,
+and `--restart-now` respectively. In follow-ups to the initial implementation,
+other methods of specifying which VMs to convert the machine types of will be
+added, such as specifying a single VM by name, or configuring a limit on the
+number of VMs that can be restarted at once when restarting the running VMs
+immediately.
+
+Given a VM with the following spec:
+```yaml
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+...
+spec:
+  running: false
+  template:
+    spec:
+      domain:
+        machine:
+          type: pc-q35-rhel8.2.0
+```
+When the MMTT is invoked, the VM spec will be updated to: 
+```yaml
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+...
+spec:
+  running: false
+  template:
+    spec:
+      domain:
+        machine:
+          type: pc-q35-rhel9.2.0
+```
+where `pc-q35-rhel9.2.0` is the latest machine type version.
+
+Along with machine types being in the format `pc-q35-rhelx.x.x`, a VM may have
```

Comment:
I am not sure all machine types are in format of  `pc-q35-rhelx.x.x`,

Reply:
Sorry, in this case in the proposal i was using the `pc-q35-rhelx.x.x` machine type as an example pertaining specifically to the CentOS Stream example I mentioned in the overview; however the actual implementation will not just be limited to machine types of this format/vendor specific machine types.

Reply:
Please leave a note outlining this detail.